commons-math/src/main/java/org/apache/commons/math4/linear/SingularValueDecomposition.java

```diff
@@ -0,0 +1,744 @@
+import org.apache.commons.math4.util.FastMath;
+import org.apache.commons.math4.util.Precision;
+
+/**
+ * Calculates the compact Singular Value Decomposition of a matrix.
+ * <p>
+ * The Singular Value Decomposition of matrix A is a set of three matrices: U,
+ * &Sigma; and V such that A = U &times; &Sigma; &times; V<sup>T</sup>. Let A be
+ * a m &times; n matrix, then U is a m &times; p orthogonal matrix, &Sigma; is a
+ * p &times; p diagonal matrix with positive or null elements, V is a p &times;
+ * n orthogonal matrix (hence V<sup>T</sup> is also orthogonal) where
+ * p=min(m,n).
+ * </p>
+ * <p>This class is similar to the class with similar name from the
+ * <a href="http://math.nist.gov/javanumerics/jama/">JAMA</a> library, with the
+ * following changes:</p>
+ * <ul>
+ *   <li>the {@code norm2} method which has been renamed as {@link #getNorm()
+ *   getNorm},</li>
+ *   <li>the {@code cond} method which has been renamed as {@link
+ *   #getConditionNumber() getConditionNumber},</li>
+ *   <li>the {@code rank} method which has been renamed as {@link #getRank()
+ *   getRank},</li>
+ *   <li>a {@link #getUT() getUT} method has been added,</li>
+ *   <li>a {@link #getVT() getVT} method has been added,</li>
+ *   <li>a {@link #getSolver() getSolver} method has been added,</li>
+ *   <li>a {@link #getCovariance(double) getCovariance} method has been added.</li>
+ * </ul>
+ * @see <a href="http://mathworld.wolfram.com/SingularValueDecomposition.html">MathWorld</a>
+ * @see <a href="http://en.wikipedia.org/wiki/Singular_value_decomposition">Wikipedia</a>
+ * @since 2.0 (changed to concrete class in 3.0)
+ */
+public class SingularValueDecomposition {
+    /** Relative threshold for small singular values. */
+    private static final double EPS = 0x1.0p-52;
+    /** Absolute threshold for small singular values. */
+    private static final double TINY = 0x1.0p-966;
+    /** Computed singular values. */
+    private final double[] singularValues;
+    /** max(row dimension, column dimension). */
+    private final int m;
+    /** min(row dimension, column dimension). */
+    private final int n;
+    /** Indicator for transposed matrix. */
+    private final boolean transposed;
+    /** Cached value of U matrix. */
+    private final RealMatrix cachedU;
+    /** Cached value of transposed U matrix. */
+    private RealMatrix cachedUt;
+    /** Cached value of S (diagonal) matrix. */
+    private RealMatrix cachedS;
+    /** Cached value of V matrix. */
+    private final RealMatrix cachedV;
+    /** Cached value of transposed V matrix. */
+    private RealMatrix cachedVt;
+    /**
+     * Tolerance value for small singular values, calculated once we have
+     * populated "singularValues".
+     **/
+    private final double tol;
+
+    /**
+     * Calculates the compact Singular Value Decomposition of the given matrix.
+     *
+     * @param matrix Matrix to decompose.
+     */
+    public SingularValueDecomposition(final RealMatrix matrix) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 83. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 112-199 into a separate method.